### PR TITLE
[FEATURE] SQLite

### PR DESCRIPTION
**Pull request by grosenberger**
This PR supports https://github.com/OpenMS/OpenMS/pull/2480.

### CODE REVIEW

**Comment by cbielow** [comment]
nice. But did you check if Qts sqlite might do the job as well?
its a dependency already, but I'm not sure of all the distros ship the sqlite drivers:
http://doc.qt.io/qt-4.8/sql-driver.html  or soon https://doc.qt.io/qt-5/qtsql-attribution-sqlite.html

**Comment by hroest** [comment]
yes, basically it would also be possible to go through Qt. It would require us to adopt a lot of the code that we have written but its not impossible. There used to SQL support in OpenMS in the early days...

**Comment by timosachsenberg** [comment]
+1 maybe hear from @jpfeuffer as this affects the contrib and test infrastructure

**Comment by hroest** [comment]
also we should check how travis will work. Installing libsqlite-dev should not be a big problem though

**Comment by timosachsenberg** [comment]
sure feel free to add it to the Travis yaml

**Comment by hroest** [comment]
can somebody add the sqlite file from here http://ftp.osuosl.org/pub/blfs/conglomeration/sqlite/ to the ftp file where the contrib will look for it? E.g. whoever has access to http://ftp.mi.fu-berlin.de/pub/OpenMS/contrib-sources/

**Comment by hroest** [comment]
it seems this is critical to get the OpenMS pull request moving as the sqlite 12.04 is not sufficient (that was painful to find out).

**Comment by jpfeuffer** [comment]
It is uploaded now.

**Comment by grosenberger** [comment]
Excellent, thank you @jpfeuffer!

**Comment by jpfeuffer** [comment]
I'll try a rebuild (I don't know why it tries all combinations -.-)

**Comment by jpfeuffer** [comment]
Ok nice. Worked on all unix variants.

**Comment by hroest** [comment]
great, so the next steps will be

- merge this PR
- update contrib and build SQLite on all nightly test machines
- merge the main PR OpenMS/OpenMS#2480

**Comment by timosachsenberg** [comment]
any reason why patches/sqlite/Makefile_in.diff is empty?

File: libraries.cmake/sqlite.cmake [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    OPENMS_BUILDLIB("SQLITE-cblas (Release)" MSBUILD_ARGS_SLN MSBUILD_ARGS_TARGET "Release" SQLITE_DIR)
+  
+    ## copy includes
+    set(dir_target ${PROJECT_BINARY_DIR}/include/sqlite)
+    set(dir_source ${SQLITE_DIR}/win32_VS${CONTRIB_MSVC_VERSION}/include/sqlite)
+    OPENMS_COPYDIR(dir_source dir_target)
+
+  else()
+    # patch Makefile.in
+    set(PATCH_FILE "${PATCH_DIR}/sqlite/Makefile_in.diff")
```
Comment (hroest) [inline libraries.cmake/sqlite.cmake]: why do we patch here with an empty file?

@grosenberger is this just a copy-paste error or is there a reason for this?

File: libraries.cmake/sqlite.cmake [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    set(ZIP_ARGS "x -y -osrc")
+  else()
+    set(ZIP_ARGS "xzf")
+  endif(MSVC)
+  OPENMS_SMARTEXTRACT(ZIP_ARGS ARCHIVE_SQLITE "SQLITE" "INSTALL")
+  
+  if(MSVC)
+    if (CONTRIB_MSVC_VERSION STREQUAL "12")
+      # patch config.h (for VS12 only, since math.h defines some functions which SQLITE re-defines otherwise)
+      set(PATCH_FILE "${PATCH_DIR}/sqlite/config.diff")
```
Comment (hroest) [inline libraries.cmake/sqlite.cmake]: @grosenberger it seems that this will fail as there is no config patch file

**Comment by hroest** [comment]
We also need to test this with a few Windows machines, I am not sure yet that the MSVS code paths will work for all versions (e.g. since a patch file is missing for example).

File: libraries.cmake/sqlite.cmake (hunk 1 above)
Comment (grosenberger) [inline libraries.cmake/sqlite.cmake]: C&P error

File: libraries.cmake/sqlite.cmake (hunk 2 above)
Comment (grosenberger) [inline libraries.cmake/sqlite.cmake]: C&P error

**Comment by grosenberger** [comment]
The makefile was only tested on macOS & Linux and we should definitely test this on Windows. Unfortunately, I don't have access to a machine, does anybody else?

**Comment by jpfeuffer** [comment]
```
CMake Error at macros.cmake:201 (message):
  Microsoft (R) Build Engine version 4.6.1055.0


  [Microsoft .NET Framework, version 4.0.30319.42000]


  Copyright (C) Microsoft Corporation.  All rights reserved.


  


  
  C:\dev\jenkins-tuebingen\ws\openms_contrib_CI\640e7a4d\contrib_build\src\CoinMP-1.8.3\CoinMP\MSVisualStudio\v10\CoinMP.sln.metaproj
  : error MSB4057: The target "sqlite" does not exist in the project.
  [C:\dev\jenkins-tuebingen\ws\openms_contrib_CI\640e7a4d\contrib_build\src\CoinMP-1.8.3\CoinMP\MSVisualStudio\v10\CoinMP.sln]
Call Stack (most recent call first):
  libraries.cmake/sqlite.cmake:17 (OPENMS_BUILDLIB)
  CMakeLists.txt:540 (OPENMS_CONTRIB_BUILD_SQLITE)


-- Configuring incomplete, errors occurred!
```

File: libraries.cmake/sqlite.cmake [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  if(MSVC)
+    set(ZIP_ARGS "x -y -osrc")
+  else()
+    set(ZIP_ARGS "xzf")
+  endif(MSVC)
+  OPENMS_SMARTEXTRACT(ZIP_ARGS ARCHIVE_SQLITE "SQLITE" "INSTALL")
+  
+  if(MSVC)
+    set(MSBUILD_ARGS_TARGET "sqlite")
+    OPENMS_BUILDLIB("SQLITE (Debug)" MSBUILD_ARGS_SLN MSBUILD_ARGS_TARGET "Debug" SQLITE_DIR)
```
Comment (hroest) [inline libraries.cmake/sqlite.cmake]: yes, this very clearly will not work. there is no solution file set, e.g. `MSBUILD_ARGS_SLN` is empty!

**Comment by hroest** [comment]
same here, but I created a branch that works: https://github.com/hroest/contrib/tree/feature/win_sqlite

if @grosenberger can pull it in, then it will be part of this PR
https://github.com/grosenberger/contrib/pull/1

**Comment by grosenberger** [comment]
Thank you very much for the patches, @hroest!

**Comment by jpfeuffer** [comment]
Hmm.. I get an error on the build machines Win8 VS12 as well as Win8 VS15:

`-- Configure SQLITE library using nmake ... in  C:/dev/jenkins-tuebingen/ws/openms_contrib_CI/640e7a4d/contrib_build/src/sqlite-autoconf-3150000
-- Configuring sqlite failed
CMake Error: File C:/dev/jenkins-tuebingen/ws/openms_contrib_CI/640e7a4d/contrib_build/src/sqlite-autoconf-3150000/sqlite3.dll does not exist.
CMake Error at libraries.cmake/sqlite.cmake:37 (configure_file):
  configure_file Problem configuring file
Call Stack (most recent call first):
  CMakeLists.txt:540 (OPENMS_CONTRIB_BUILD_SQLITE)


CMake Error: File C:/dev/jenkins-tuebingen/ws/openms_contrib_CI/640e7a4d/contrib_build/src/sqlite-autoconf-3150000/sqlite3.lib does not exist.
CMake Error at libraries.cmake/sqlite.cmake:38 (configure_file):
  configure_file Problem configuring file
Call Stack (most recent call first):
  CMakeLists.txt:540 (OPENMS_CONTRIB_BUILD_SQLITE)`

**Comment by jpfeuffer** [comment]
More logs:
```
find: '#define SQLITE_VERSION ': No such file or directory
The process tried to write to a nonexistent pipe.
NMAKE : fatal error U1077: '"c:\Program Files (x86)\Microsoft SDKs\Windows\v7.0A\bin\x64\rc.EXE"' : return code '0x1'
Stop.
	cl -nologo -W3   -DSQLITE_OS_WIN=1 -I. -I. -fp:precise -MT -DNDEBUG -D_CRT_SECURE_NO_DEPRECATE -D_CRT_SECURE_NO_WARNINGS -D_CRT_NONSTDC_NO_DEPRECATE -D_CRT_NONSTDC_NO_WARNINGS -DSQLITE_THREADSAFE=1 -DSQLITE_THREAD_OVERRIDE_LOCK=-1 -DSQLITE_TEMP_STORE=1  -DSQLITE_MAX_TRIGGER_DEPTH=100  -DSQLITE_ENABLE_FTS3=1 -DSQLITE_ENABLE_RTREE=1 -DSQLITE_ENABLE_COLUMN_METADATA=1   -O2 -Zi -Fosqlite3.lo  -c sqlite3.c
sqlite3.c
	echo #ifndef SQLITE_RESOURCE_VERSION > sqlite3rc.h
	echo #define SQLITE_RESOURCE_VERSION  >> sqlite3rc.h
	echo #endif >> sqlite3rc.h
	rc -DSQLITE_OS_WIN=1 -I. -I.   -DNDEBUG -D_CRT_SECURE_NO_DEPRECATE -D_CRT_SECURE_NO_WARNINGS -D_CRT_NONSTDC_NO_DEPRECATE -D_CRT_NONSTDC_NO_WARNINGS -DSQLITE_THREADSAFE=1 -DSQLITE_THREAD_OVERRIDE_LOCK=-1 -DSQLITE_TEMP_STORE=1  -DSQLITE_MAX_TRIGGER_DEPTH=100  -DSQLITE_ENABLE_FTS3=1 -DSQLITE_ENABLE_RTREE=1 -DSQLITE_ENABLE_COLUMN_METADATA=1   -r -fo sqlite3res.lo -DRC_VERONLY .\sqlite3.rc
Microsoft (R) Windows (R) Resource Compiler Version 6.1.7600.16385

Copyright (C) Microsoft Corporation.  All rights reserved.

.\sqlite3.rc(54) : error RC2127 : version WORDs separated by commas expected
```

**Comment by jpfeuffer** [comment]
Okay. Works if I login and build using the VS command line.
Probably related to the fact that we use Git Bash to start the VS command line (to have the environment correctly setup) which then starts cmake

**Comment by hroest** [comment]
ok, great so it seems that it builds on all platforms and is installed on the nighly build servers. Should me merge?

**Pull request merged**